### ISSUE TITLE
In Ansible version 2.3.2 the S3 module refuses to do checksum on multipart uploads

### ISSUE THREAD

**Issue by einarc** [ansible/ansible#28744]
##### ISSUE TYPE

 - Bug Report
 

##### ANSIBLE VERSION

```
ansible 2.3.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Aug  2 2016, 04:20:16) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

##### SUMMARY

Ansible CORE 2.3.2 S3 module refuses to do checksum on multipart uploads whereas 2.2.1 worked perfectly, why was the behavior changed in such a drastic way? Is this a bug?

See problems reported here:
https://access.redhat.com/support/cases/#/case/01904604

##### STEPS TO REPRODUCE

Upload a file to s3 that requires multiparts then attempt to download it using the s3 module.

```yaml
- name: Inject SQL - Get dump
  s3: bucket="{{ s3_bucket }}" object="{{ s3_key }}" dest="{{ s3_sql_file }}" mode=get overwrite=always
  when: create_rds
  tags: rds_database
```

##### ACTUAL RESULTS

The s3 module errors out saying it cannot apply checksum logic to the multi-part s3 file it's trying to download. Notice that we downsized the file from 2 GB to 630 MB and still got the same error. We don't understand why this gets triggered in the newer version but it doesn't in 2.2.1

```
{
    "_ansible_parsed": true,
    "_ansible_no_log": false,
    "_ansible_delegated_vars": {
        "ansible_delegated_host": "localhost",
        "ansible_host": "localhost"
    },
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": ""
    },
    "msg": "Files uploaded with multipart of s3 are not supported with checksum, unable to compute checksum."
}
```
This was triggered trying to solve another issue, which indeed 2.3.2 solved, but I cannot upgrade from 2.2.1 as long as this one is outstanding.
https://github.com/ansible/ansible/issues/28087
[2.3.2-Ansible-Core_Failed (1).txt](https://github.com/ansible/ansible/files/1261422/2.3.2-Ansible-Core_Failed.1.txt)
[2.2.1-Ansible-Core-Inject-DB-Success (1).txt](https://github.com/ansible/ansible/files/1261423/2.2.1-Ansible-Core-Inject-DB-Success.1.txt)

**Comment by willthames** [comment]
Looks like this is already fixed in the `devel` branch:

https://github.com/ansible/ansible/commit/1de91a9aa0#diff-68ba194b6312d1232756a1ecb586474bL309

The question remains whether we want to backport just the multipart fix to 2.3 (we can't just cherry pick the commit as it's too big for a stable branch)

**Comment by Tombar** [comment]
This is affecting us also on 2.3.2.0, is there any workaround?

**Comment by cneilwarner** [comment]
And us. Is there a workaround?

**Comment by einarc** [comment]
@carawarner  There is, use aws cli and forget about this. Just do shell: aws s3 cp src dest.

You'll need:
1.-Access to the AWS s3 api with either role or passing credentials to the awx user via environment variables:
If you have Tower it's easy, the Keys are set at run time when you run a job. If you are running ansible from console then I'd suggest you use vault.
Here's an example for Tower, add these to the bottom of your task/include.
environment:
   AWS_ACCESS_KEY_ID:  "{{ lookup('env','AWS_ACCESS_KEY') }}"
   AWS_SECRET_ACCESS_KEY:  "{{ lookup('env','AWS_SECRET_KEY') }}"

The task/include will copy the environment variables from the ansible master and set them at run time on the python client then run the shell command. Where you're running the s3 cp command you must install aws cli.

**Comment by willthames** [comment]
Workaround - you should be able to use the latest aws_s3 module with 2.3.2

https://willthames.github.io/2017/12/12/using-updated-modules-with-stable-ansible.html

**Comment by willthames** [comment]
Closing this issue as 2.3.2 is now only in critical bug fix mode and the issue does not occur in latest 2.4 releases onwards.

https://docs.ansible.com/ansible/2.5/reference_appendices/release_and_maintenance.html

**Comment by einarc** [comment]
@willthames  That statement is innacurate, I just had to change another call into aws_cli and I'm using exactly ansible 2.3.2

TBH the nightmare with upgrading ansible is that either the way things get done are deprecated (which is actually a case where we can do something about it, but it happens rather quickly regardless) or worse, new bugs get introduced...and there's nothing we can do about it but rollback, or wait again. I've been in 2.3.2 for over a year now because of those two reasons, mainly the second. This particular bug triggered when files reach a certain limit. In my case a database file stored in S3 grew. In summary I don't want to break more things just to get this fixed....filing tickets against RH who bought ansible does not make things speedier necessarily.

For the record I've been with you guys since 1.9., full time. So where all this lifecycle leaves us then?

[ec2-user@ip-172-23-1-229 ~]$ ansible --version
ansible 2.3.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Aug  2 2016, 04:20:16) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
[ec2-user@ip-172-23-1-229 ~]$ date
Wed Apr  4 20:27:59 EDT 2018

**Comment by willthames** [comment]
Sorry, which statement is inaccurate? Are you saying this issue still happens in 2.4.4 or devel?

You're right about problems being introduced. Let me tell you what I've personally done about. I've created [policies](https://github.com/ansible/community/blob/master/group-aws/integration.md#policy) that:
* all feature requests for stableinterface modules must contain test cases
* all feature requests for preview modules should contain test cases
* all new modules must contain test cases

I can't fix the past, and I'd urge you to look at 2.5 again. If you can't move to 2.5, there are ways to use 2.5 modules with 2.3 core as I [mention above](https://github.com/ansible/ansible/issues/28744#issuecomment-378456342). 

If you're still having problems with that, just let me know here and I'll see what else can be done.
